### PR TITLE
TPUSpawn + IterableDataset error message

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Adds a more firendly error message when using iterable dataset + TPUSpawn (currently this fails inside torch XLA here: https://github.com/pytorch/xla/blob/86463f894eb21f31dcd3449e4e55cc45b8a8586a/torch_xla/distributed/parallel_loader.py#L89)

See example outputs here: https://colab.research.google.com/drive/1fXtIfIemyzuF2Flxvyjd0aHz6zL8u5AE?usp=sharing

Would be nice if this could happen on just one rank (perhaps switching to rank_zero_warn?), can't happen pre-accelerator as needs the dataloaders to actually exist

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs) PL Slack
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [N/A] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [N/A] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 1]
```diff
@@ -64,7 +65,11 @@ def distributed_sampler_kwargs(self) -> dict:
     def is_distributed(self):
         return self.world_size != 1
 
-    def process_dataloader(self, dataloader: Union[Iterable, torch.utils.data.DataLoader]) -> MpDeviceLoader:
+    def process_dataloader(self, dataloader: torch.utils.data.DataLoader) -> MpDeviceLoader:
+        if not has_len(dataloader):
+            raise MisconfigurationException(
+                "TPUSpawn does not currently support IterableDataset objects, the dataset must implement __len__."
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:71]:
```suggestion
                "TPUs does not currently support IterableDataset objects, the dataset must implement __len__."
                "HINT: You can mock the length on your dataset to bypass this MisconfigurationException"
```

**Review by tchaton — APPROVED** [review]

**Comment by ethanwharris** [comment]
@tchaton Thanks for the review :smiley: I've added an earlier message when the dataloaders are passed to fit and some tests, mind re-reviewing?

**Review by kaushikb11 — APPROVED** [review]
LGTM!

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 2]
```diff
@@ -40,13 +42,50 @@
...
+
+        for dataloader in dataloaders:
+            if not has_len(dataloader):
+                raise MisconfigurationException(
+                    "TPUs do not currently support IterableDataset objects, the dataset must implement __len__."
+                    " HINT: You can mock the length on your dataset to bypass this MisconfigurationException."
+                )
+
+    @staticmethod
+    def _validate_patched_dataloaders(model: 'Module') -> None:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:70]: Shouldnt the annotation be `LightningModule`? As we are accessing its attributes `*_dataloader`

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 3]
```diff
@@ -40,13 +42,50 @@
...
 
+    @staticmethod
+    def _validate_dataloader(dataloaders: Union[List['DataLoader'], 'DataLoader']):
+        if not isinstance(dataloaders, list):
+            dataloaders = [dataloaders]
+
+        for dataloader in dataloaders:
+            if not has_len(dataloader):
+                raise MisconfigurationException(
+                    "TPUs do not currently support IterableDataset objects, the dataset must implement __len__."
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:65]:
```suggestion
                    "TPUs do not currently support IterableDataset objects, the dataset must implement `__len__`."
```

File: tests/plugins/test_tpu_spawn.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    [
+        (_loader_no_len, None, None, None),
+        (None, _loader_no_len, None, None),
+        (None, None, _loader_no_len, None),
+        (None, None, None, _loader_no_len),
+        (None, [_loader, _loader_no_len], None, None),
+    ],
+)
+def test_error_patched_iterable_dataloaders(
+        tmpdir, train_dataloader, val_dataloaders, test_dataloaders, predict_dataloaders
```
Comment (carmocca) [inline tests/plugins/test_tpu_spawn.py:55]:
```suggestion
    tmpdir, train_dataloader, val_dataloaders, test_dataloaders, predict_dataloaders
```

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 2 above)
Comment (ethanwharris) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:70]: Good point! I've opted to instead use hasattr checks before trying to access `*_dataloader` stuff

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 5]
```diff
@@ -40,13 +42,50 @@
...
+            TPUSpawnPlugin._validate_dataloader(model.test_dataloader.dataloader)
+
+        if hasattr(model, 'predict_dataloader') and isinstance(model.predict_dataloader, _PatchDataLoader):
+            TPUSpawnPlugin._validate_dataloader(model.predict_dataloader.dataloader)
+
+    def connect(self, model: 'Module') -> None:
+        TPUSpawnPlugin._validate_patched_dataloaders(model)
+        return super().connect(model)
+
     def setup(self, model: torch.nn.Module) -> torch.nn.Module:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:89]:
```suggestion
    def setup(self, model: 'Module') -> 'Module':
```

**Review by carmocca — APPROVED** [review]
